slides/08_shiny.qmd

```diff
@@ -0,0 +1,442 @@
+      need(yvar %in% names(df), glue("yvar \"{yvar}\" does not exist in data")),
+      need(nrow(df) > 3, "too few data points for meaningful plot")
+    )
+    
+    # write "business logic" in a unit testable function
+    my_special_plot(df[[xvar]], df[[yvar]])
+  })
+}
+```
+
+## Reactivity: Debugging with global variables
+
+```{r, echo = TRUE}
+#| code-line-numbers: "16"
+server <- function(input, output, session) {
+  output$plot <- renderPlot({
+    # resolve reactive values
+    xvar <- input$xvar
+    yvar <- input$yvar
+    req(xvar, yvar)
+    
+    # validate – is it likely that the plot will be meaningful?
+    validate(
+      need(xvar %in% names(df), glue("xvar \"{xvar}\" does not exist in data")),
+      need(yvar %in% names(df), glue("yvar \"{yvar}\" does not exist in data")),
+      need(nrow(df) > 3, "too few data points for meaningful plot")
+    )
+    
+    # write "business logic" in a unit testable function
+    .GlobalEnv$xvar <- xvar; .GlobalEnv$yvar <- yvar;  stop()
+    my_special_plot(df[[xvar]], df[[yvar]])
+  })
+}
+```
+
+## Reactivity: "Stringy" reactivity graph
+
+::: columns
+::: {.column width="50%"}
+Bad
+![](resources/connected.png){fig-align="center"}
+Many edges, difficult to anticipate reactivity behaviour
+:::
+ 
+::: {.column width="50%"}
+Good
+![](resources/stringy.png){fig-align="center"}
+This "stringy" reactivity graph is easier to understand and debug
+:::
+:::
+
+# Testing
+
+## Testing: From Icecream to Pyramid
+
+::: columns
+::: {.column width="50%"}
+Bad
+
+![](resources/icecream.png){height=400}
+
+- cannot always test manually
+- can easily introduce bugs
+:::
+ 
+::: {.column width="50%"}
+Good
+
+![](resources/pyramid.png){height=400}
+
+- manual testing is reduced
+- only for "playing around"
+:::
+:::
+
+## Testing: Snapshot tests for UI
+
+::: columns
+::: {.column width="50%"}
+```{r, echo = TRUE, eval = FALSE}
+#| code-line-numbers: "5-9"
+test_that("myInput UI works", {
+  input <- "foo"
+  set.seed(123)
+  datasets <- mock_datasets()
+  expect_snapshot(myInput(
+    "my_test",
+    datasets = datasets,
+    input = input
+  ))
+})
+```
+:::
+
+::: {.column width="50%"}
+- Shiny UI functions return shiny tag objects, i.e. HTML code
+- Can use snapshot tests to avoid accidental changes
+- Value is limited
+- Can create issues sometimes (non-reproducible hashes)
+:::
+:::
+
+## Testing: Shiny server tests
+
+::: columns
+::: {.column width="50%"}
+```{r, echo = TRUE, eval = FALSE}
+test_that("server works", {
+  testServer(server, {
+    session$setInputs(...)
+
+    print(reactive1())
+    print(output$output1)
+    # etc
+
+    # To interactively play:
+    # browser()
+  
+    expect_equal(...)
+    # etc.
+  })
+})
+
+```
+:::
+
+::: {.column width="50%"}
+- Code is run inside the server function
+- Can access reactives, outputs, etc.
+- Session object simulates user actions and time
+- Note: need to insert `browser()`
+- Also works for module server functions
```

Comment:
and shinyApp object or app path; this can be helpful to test the shiny directly?

Reply:
if shinyApp can be used, then Javascript could be still used right?

Reply:
this comes below, testServer is different

Reply:
my bad. `testServer` works for both shiny app object,  a path to shiny app folder, module server, or server functions. But of cource it do not have front end and js could not be executed. it only creates a dummy session that can interact with the server function extracted from the shiny app object. A good thing for `shiny app object` to be used is that it could load some `onStart` functions to be executed, which is more similar to the actual app run (but only testing the server side)

Reply:
ok thanks - so we can keep as is